hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/key/KeyComparator.java

```diff
@@ -0,0 +1,212 @@
+/*
+ * Copyright (C) 2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.spi.key;
+
+import com.hedera.hapi.node.base.ContractID;
+import com.hedera.hapi.node.base.Key;
+import com.hedera.hapi.node.base.Key.KeyOneOfType;
+import com.hedera.hapi.node.base.KeyList;
+import com.hedera.hapi.node.base.ThresholdKey;
+import com.hedera.pbj.runtime.io.buffer.Bytes;
+import java.util.Comparator;
+import java.util.List;
+
+/**
+ * Comparator used to impose a deterministic ordering on collections of Key objects.
+ * <br>These include maps, sets, lists, arrays, etc...
+ * <br>The methods in this class are used in hot spot code, so allocation must be kept to a bare
+ * minimum, and anything likely to have performance questions should be avoided.
+ * <br>Note that comparing keys is unavoidably costly.  We try to exit as early as possible throughout
+ * this class, but worst case we're comparing every simple key byte-by-byte for the entire tree, which
+ * may be up to 15 levels deep with any number of keys per level.  We haven't seen a key with
+ * several million "simple" keys included, but that does not mean nobody will create one.
+ */
+public class KeyComparator implements Comparator<Key> {
```

Comment:
A thorough unit test for this class would be helpful

Reply:
True.  I'll add that to the todo list.  Might not be the next PR (that needs to get at least the majority of execution tests working), but definitely the one after that.